acme/authority.go

```diff
@@ -254,32 +273,187 @@ func (a *Authority) GetAuthz(p provisioner.Interface, accID, authzID string) (*A
 	return az.toACME(a.db, a.dir, p)
 }
 
-// ValidateChallenge attempts to validate the challenge.
+// The challenge validation state machine looks like:
+//
+// * https://tools.ietf.org/html/rfc8555#section-7.1.6
+//
+// While in the processing state, the server may retry as it sees fit. The challenge validation strategy
+// needs to be rather specific in order for retries to work in a replicated, crash-proof deployment.
+// In general, the goal is to allow requests to hit arbitrary instances of step-ca while managing retry
+// responsibility such that multiple instances agree on an owner. Additionally, when a deployment of the
+// CA is in progress, the ownership should be carried forward and new, updated (or in general, restarted),
+// instances should pick back up where the crashed instance left off.
+//
+// The steps are:
+//
+// 1. Upon incoming request to the challenge endpoint, take ownership of the retry responsibility.
+//  (a) Set Retry.Owner to this instance's ordinal (STEP_CA_ORDINAL).
+//  (b) Set Retry.NumAttempts to 0 and Retry.MaxAttempts to the desired max.
+//  (c) Set Challenge.Status to "processing"
+//  (d) Set retry_after to a time (retryInterval) in the future.
+// 2. Perform the validation attempt.
+// 3. If the validation attempt results in a challenge that is still processing, schedule a retry.
+//
+// It's possible that another request to re-attempt the challenge comes in while a retry attempt is
+// pending from a previous request. In general, these old attempts will see that Retry.NextAttempt
+// is in the future and drop their task. But this also might have happened on another instance, etc.
```

Comment:
Mention that if it happens on another instance, the ordinal will have changed or just omit.